crates/shared/src/sources/balancer_v2/graph_api.rs

```diff
@@ -153,26 +118,80 @@ pub struct RegisteredPools {
     pub stable_pools_by_factory: HashMap<H160, Vec<RegisteredStablePool>>,
 }
 
+impl RegisteredPools {
+    fn from_pool_data(fetched_block_number: u64, pools: Vec<PoolData>) -> Result<Self> {
+        let mut registered_pools = Self {
+            fetched_block_number,
+            weighted_pools_by_factory: Default::default(),
+            stable_pools_by_factory: Default::default(),
+        };
+
+        for pool in pools {
+            let common = CommonPoolData {
+                pool_id: pool.id,
+                pool_address: pool.address,
+                tokens: pool.tokens.iter().map(|token| token.address).collect(),
+                scaling_exponents: pool
+                    .tokens
+                    .iter()
+                    .map(|token| scaling_exponent_from_decimals(token.decimals))
+                    .collect::<Result<_>>()?,
+                block_created: fetched_block_number,
+            };
+            match pool.pool_type {
+                PoolType::Weighted => registered_pools
+                    .weighted_pools_by_factory
+                    .entry(pool.factory)
+                    .or_default()
+                    .push(RegisteredWeightedPool {
+                        common,
+                        normalized_weights: pool
+                            .tokens
+                            .iter()
+                            .map(|token| {
+                                token.weight.ok_or_else(|| {
+                                    anyhow!("missing weights for pool {:?}", pool.id)
+                                })
+                            })
+                            .collect::<Result<_>>()?,
+                    }),
+                PoolType::Stable => registered_pools
+                    .stable_pools_by_factory
+                    .entry(pool.factory)
+                    .or_default()
+                    .push(RegisteredStablePool { common }),
+            }
+        }
+
+        Ok(registered_pools)
+    }
+}
+
+fn scaling_exponent_from_decimals(decimals: u8) -> Result<u8> {
```

Comment:
I thought I saw this in a different file in the indexing refactor. Have we duplicated it or is it just moving around a lot?

Reply:
It moved.